src/main/scala/workflow/WorkflowUtils.scala

```diff
@@ -0,0 +1,124 @@
+package workflow
+
+import scala.collection.mutable.ArrayBuffer
+
+
+object WorkflowUtils {
+  def instructionsToPipeline[A, B](instructions: Seq[Instruction]): Pipeline[A, B] = {
+    val nodes = new ArrayBuffer[Node]()
+    val dataDeps = new ArrayBuffer[Seq[Int]]()
+    val fitDeps = new ArrayBuffer[Option[Int]]()
+    val instructionIdToNodeId = scala.collection.mutable.Map.empty[Int, Int]
+    instructionIdToNodeId.put(Pipeline.SOURCE, Pipeline.SOURCE)
+
+    for (instruction <- 0 until instructions.length) {
+      instructions(instruction) match {
+        case est: EstimatorNode => Unit
+        case transformer: TransformerNode => Unit
+        case source: SourceNode => {
+          instructionIdToNodeId.put(instruction, nodes.length)
+          nodes.append(source)
+          dataDeps.append(Seq())
+          fitDeps.append(None)
+        }
+        case TransformerApplyNode(transformer, inputs) => {
+          instructions(transformer) match {
+            case transformerNode: TransformerNode => {
+              instructionIdToNodeId.put(instruction, nodes.length)
+              nodes.append(transformerNode)
+              dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+              fitDeps.append(None)
+            }
+            case estimatorFitNode: EstimatorFitNode => {
+              instructionIdToNodeId.put(instruction, nodes.length)
+              // TODO: Get more reasonable label...
+              nodes.append(new DelegatingTransformerNode("Estimator fit"))
+              dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+              fitDeps.append(Some(instructionIdToNodeId(transformer)))
+            }
+            case _ => throw new RuntimeException("Transformer apply instruction must point at a Transformer")
+          }
+        }
+        case EstimatorFitNode(est, inputs) => {
+          val estimatorNode = instructions(est).asInstanceOf[EstimatorNode]
+          instructionIdToNodeId.put(instruction, nodes.length)
+          nodes.append(estimatorNode)
+          dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+          fitDeps.append(None)
+        }
+      }
+    }
+
+    new ConcretePipeline(nodes.toSeq, dataDeps.toSeq, fitDeps.toSeq, nodes.length - 1)
+  }
+
+  def pipelineToInstructions[A, B](pipeline: Pipeline[A, B]): Seq[Instruction] = {
+    val nodes = pipeline.nodes
+    val dataDeps = pipeline.dataDeps
+    val fitDeps = pipeline.fitDeps
+    val sink = pipeline.sink
+
+    pipelineRecurse(sink, nodes, dataDeps, fitDeps, Map(Pipeline.SOURCE -> Pipeline.SOURCE), Seq())._2
+  }
+
+  def pipelineRecurse(
```

Comment:
It would be nice to summarize how this works in a scaladoc.

Something like "Linearizes the pipeline DAG by by walking backwards from the sink in a depth-first manner."